docs/understanding-your-pipeline/canonical-event/index.md

```diff
@@ -36,274 +42,490 @@ The Snowplow data table is designed to be immutable: the data in each line shoul
+...
+from 
+    atomic.events
+```
+
+</TabItem>
+<TabItem value="db" label="Databricks">
+
+The unstructured event is not part of the standard `atomic.events` columns, it is instead a `struct` type column with the name of the schema and the major version number. The fields in this table will be determined by the JSON schema defined for the event in advance, converted to snake case. Users can query fields in the self-describing event by extracting them using 
 
-When a user sends in a custom unstructured event, they do so as a JSON of name-value properties that conforms to a JSON schema defined for the event earlier.
+```sql
+select
+...
+unstruct_event_my_example_event_1_0_0.myCustomField,
+...
+from 
+    atomic.events
+```
+
+</TabItem>
+</Tabs>
+
+
+#### Entities (Contexts)
 
-The unstructured event is not part of the `atomic.events` table; instead, for users running on Redshift, it is shredded into its own table. The fields in this table will be determined by the JSON schema defined for the event in advance. Users can query just the table for that particular unstructured event, if that's all that's required for their analysis, or join that table back to the `atomic.events` table by
+Entities (Contexts) enable Snowplow users to define their own entities that are related to events, and fields that are related to each of those entities. For example, an online retailer may choose to define a `user` context, to store information about a particular user, which might include data points like the users Facebook ID, age, membership details etc. In addition, they may also define a `product` context, with product data e.g. SKU, name, created date, description, tags etc.
+
+An event can have any number of custom entities attached. Each context is passed into Snowplow as a JSON. Additionally, the [Snowplow Enrichment](/docs/enriching-your-data/index.md) process can derive additional contexts.
+
+<Tabs groupId="warehouse" queryString>
+<TabItem value="rs/pg" label="Redshift/Postgres" default>
+
+Contexts are not part of the `atomic.events` table; instead, for users running on Redshift, Snowplow will shred each context JSON into a dedicated table in the `atomic` schema, making it much more efficient for analysts to query data passed in in any one of the contexts. Those contexts can be joined back to the core `atomic.events` table by the following, which is a one-to-one join (for a single record entity) or a one-to-many join (for a multi-record entity), assuming no duplicates.
 
 ```sql
-atomic.my_example_unstructured_event_table.root_id = atomic.events.event_id
+select 
+    ...
+from 
+    atomic.events ev
+left join 
+    atomic.my_custom_context_table ctx
+    on ctx.root_id = ev.event_id and ctx.root_tstamp = ev.collector_tstamp
 ```
 
-#### Contexts
+:::caution
+
+You may need to take care of duplicate events as Snowplow uses an _at least once_ method of delivery.
+
+<details>
+<summary>How to join your data with duplicates?</summary>
```

Comment:
This is valuable content, probably warrants its own page or something, though?

Reply:
Maybe could stick it on https://docs.snowplow.io/docs/destinations/warehouses-and-lakes/rdb/transforming-enriched-data/deduplication/ ?

Reply:
Seems more discoverable but at the same time it's kind of the opposite of that page, "how do dedupe when all this stuff hasn't worked".
It's fine by me as is on this page so we'll wait and see if someone else complains. :stuck_out_tongue: